### ISSUE TITLE
support for on-premise OCP

### ISSUE THREAD

**Issue by derek82511** [openshift/windows-machine-config-operator#141]
Can this Operator support for on-premise OCP 4.x ?
Or it is in the roadmap?

**Comment by aravindhp** [comment]
@derek82511, this is on our roadmap.

**Comment by jxadro** [comment]
Any news?

I think this operator is currently GA, in OpenShift console, in the operator hub tab it is under "Provided by RedHat", not under community.

In the roadmap it was in doubt if VMWare was going to be supported for windows nodes for the GA.

In the documentation it is only listed AWS and Azure.

Are windows nodes created on VMWare or Baremetal supported?

Thank you.

**Comment by LorbusChris** [comment]
vSphere and Baremetal support are both on the roadmap, but not available yet.

Progress on vSphere support can be tracked here: https://github.com/openshift/windows-machine-config-operator/pull/149 (it's currently blocked on a Windows issue that'll have to be resolved from the Microsoft side)

**Comment by jxadro** [comment]
Hi @LorbusChris, if I'm not wrong the PR has been currently merged.

What is the plan to move VMWare support for Windows Nodes to GA?

Thank you.

**Comment by aravindhp** [comment]
@jxadro the plan is to release vSphere IPI support in WMCO in the 2.x version which will be released after OpenShift 4.7.

**Comment by aravindhp** [comment]
Opened an OpenShift enhancement proposal to add support for `Bring Your Own Windows Host`: https://github.com/openshift/enhancements/pull/608

**Comment by jxadro** [comment]
Hi @aravindhp 

In the current documentation I read: 
https://github.com/openshift/windows-machine-config-operator
"OKD/OCP 4.6 cluster running on Azure, AWS or vSphere configured with hybrid OVN Kubernetes networking"

With that prerequisite I understand that it is currently possible to deploy windows nodes in vsphere, but in your previous post you said that it will be in WMCO 2.x.

Please could you help to clarify the current status about supporting windows nodes on vmware?

Now that OCP 4.7 has been released, do you have any date about GA of WMCO 2.x?

Thank you.

**Comment by jxadro** [comment]
I have just installed a 4.7 OpenShift and I only see the community operator 2.0. So I guess the situation is that the windows operator currently support windows on vsphere but it is not GA yet.

Is this correct?

**Comment by aravindhp** [comment]
Hi @jxadro,

>In the current documentation I read:
>https://github.com/openshift/windows-machine-config-operator
>"OKD/OCP 4.6 cluster running on Azure, AWS or vSphere configured with hybrid OVN Kubernetes networking"

That should be fixed to say `OKD/OCP cluster running on Azure (4.6+), AWS (4.6+) or vSphere (4.7+) configured with hybrid OVN Kubernetes networking`.

>Now that OCP 4.7 has been released, do you have any date about GA of WMCO 2.x?

It has been scheduled for this week.

>So I guess the situation is that the windows operator currently support windows on vsphere but it is not GA yet.

That is correct. To be clear the support is for adding Windows nodes using MachineSets on installer provisioned clusters on vSphere and will not support adding your own Windows hosts.

**Comment by jxadro** [comment]
Thank you.

**Comment by aravindhp** [comment]
@jxadro, WMCO 2.0 has been released: [OpenShift 4.7: Windows Container Support for Red Hat OpenShift on vSphere
](https://www.openshift.com/blog/windows-container-support-for-red-hat-openshift-on-vsphere)

**Comment by jandradap** [comment]
> @jxadro, WMCO 2.0 has been released: [OpenShift 4.7: Windows Container Support for Red Hat OpenShift on vSphere ](https://www.openshift.com/blog/windows-container-support-for-red-hat-openshift-on-vsphere)

vSphere IPI.

Is vSphere UPI on the roadmap? @aravindhp

**Comment by aravindhp** [comment]
> Is vSphere UPI on the roadmap? @aravindhp

@jandradap yes it is. Please see [bring-your-own-windows-host.md](https://github.com/openshift/enhancements/blob/master/enhancements/windows-containers/bring-your-own-windows-host.md)

**Comment by jomeier** [comment]
Hi, we think about creating an on premises OpenShift cluster with UPI in an air gapped environment to support windows containers to our business.

We also use MachineSets for vSphere to create our worker nodes. We would like to do the same for the windows workers.

Problem: the docs say that ...
- https://docs.openshift.com/container-platform/4.7/windows_containers/windows-containers-release-notes-2-x.html#wmco-2-0-0: Running Windows container workloads is not supported for clusters in a restricted network or disconnected environment.
Is that true? Why? Or is this comment already outdated with newer WMCO versions?
- UPI currently seems not to be supported. During our UPI installation we only install the linux masters. All of the workers are created with MachineSets. As far as I understand that's the same with IPI installers. Is creating windows workers with MachineSets supported with UPI OpenShit clusters?
- Is Namespace isolation supported with mixed linux and windows workers?
- What is this VXLan port used for that's mentioned in the docs? https://docs.openshift.com/container-platform/4.7/windows_containers/windows-containers-release-notes-2-x.html

Thanks for your assistance.

Greetings,

Josef

**Comment by jomeier** [comment]
I don't see any evidence in the WMCO source code why air gapped environments shouldn't work. The only reason why I could imagine that this is a problem is, that previously files had to be downloaded by the Windows Machine Bootstrapper. Currently this files seem to be included in WMCO's payload:
- cni
- hybrid-overlay-node.exe
- kube-node
- powershell
- windows_exporter.exe
- wmcb.exe

If airgapped installations are now supported, that should be corrected in the docs IMHO.

**Comment by sebsoto** [comment]
Hi @jomeier at the moment air gapped environments will not work, as the pause image for all Windows pods is set to the image officially provided by microsoft. This is reflected in the [WMCB code](https://github.com/openshift/windows-machine-config-bootstrapper/blob/0d5d2f4fce523dfb6978b78c64705c9a123836ea/pkg/bootstrapper/bootstrapper.go#L44-L45), which handles the starting/stopping of kubelet.

This will be a limitation for the near future, as we do not currently provide a way to customize kubelet flags.

**Comment by jomeier** [comment]
@sebsoto:
Thanks for your answer. Because we use a pull through cache for docker images where our cluster can get images from the internet, do you think that this might work in that case?

**Comment by sebsoto** [comment]
@jomeier I can't say with certainty, but as long as you can pull `mcr.microsoft.com/oss/kubernetes/pause:3.4.1` from the Windows instance you are trying to join to the cluster it should work.

**Comment by sebsoto** [comment]
@jomeier  I say "should", as I don't know for sure. Github issues are us answering at best effort :)
If you have a subscription please open a support case through the Red Hat Customer Portal and you can get more definite answers. Off the top of my head the pause image is the main culprit for preventing air-gapped clusters, but there may be another blocker that I am not thinking of.

**Comment by jomeier** [comment]
Thank you!

**Comment by sebsoto** [comment]
@jomeier  Circling back here to answer the other questions:

> UPI currently seems not to be supported. During our UPI installation we only install the linux masters. All of the workers are created with MachineSets. As far as I understand that's the same with IPI installers. Is creating windows workers with MachineSets supported with UPI OpenShit clusters?

Based on the wording that we put out, it is not officially supported, however as long as there is a Windows Machine (created through a MachineSet), WMCO shouldn't have an issue creating a Windows node. The intent behind not supporting UPI installations, is that we currently only support Windows nodes created through Windows MachineSets, that will change in WMCO release 3.1.0 for OpenShift 4.8, coming soon.

> Is Namespace isolation supported with mixed linux and windows workers?

I'm not fully sure what scenario you are asking about

> What is this VXLan port used for that's mentioned in the docs? https://docs.openshift.com/container-platform/4.7/windows_containers/windows-containers-release-notes-2-x.html

[I'll defer to the vmware docs for that](https://docs.vmware.com/en/VMware-Validated-Design/5.0/com.vmware.vvd.sddc-design.doc/GUID-35A129F2-BC5C-49C7-8F6E-2886099ABFA4.html)

**Comment by jomeier** [comment]
That's good news. Thank's a lot, @sebsoto.
